### PR TITLE
Filter feature types

### PR DESCRIPTION
**Pull request by TheCovenant**

### CODE REVIEW

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 1]
```diff
@@ -21,25 +21,46 @@ import {getFeaturesByState} from "../../selectors/getFeaturesByState";
 import {MapFeature} from "../../states/map/MapFeature";
 import {addFilter} from "../../actions/map/AddFilterAction";
 
+const limit = 500;
 var wkt = require("terraformer-wkt-parser");
+var totalDataCounter = 0; //counter of number of features queried so far during pagination. Used for offset
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:26]: These all needs to be in the Redux state..

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 2]
```diff
@@ -21,25 +21,46 @@ import {getFeaturesByState} from "../../selectors/getFeaturesByState";
...
   >(featuresQueryDocument, {variables: {query: {types: [FeatureType.State]}}});
 
-  const [getFeaturesWithin] = useLazyQuery<
+  const [getFeaturesWithin, {loading}] = useLazyQuery<
     MapFeaturesQuery,
     MapFeaturesQueryVariables
   >(featuresQueryDocument, {
     onCompleted: (data: MapFeaturesQuery) => {
+      totalDataCounter += data.features.length;
+      latestDataCounter = data.features.length;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:62]: I have no idea why you would call this a counter.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 3]
```diff
@@ -97,68 +120,138 @@ const MapImpl: React.FunctionComponent = () => {
...
+                  };
+                }),
+              }),
+              info: {
+                id: featureType,
+              },
+            };
+
+            //dispatch(removeDataset(featuresInState[0].type));
+            dispatch(removeDataset(featureType));
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:153]: The issue said you don't need this, that addDataToMap will overwrite.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 4]
```diff
@@ -97,68 +120,138 @@ const MapImpl: React.FunctionComponent = () => {
       case MapFeatureState.LOADED: {
         /*
         Features in this state have been queried, created and pushed into the
-        features state list that is located on the store. The next step is to
-        add the data of the features to the map using the addDataToMap action which is reduced by
-        keplerGL reducer.. The geometry of the feature is used
-        to display its location and shape on the map.
+        features state list that is located on the store. They have also been pushed into the appropriate place of the
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:123]: Now that I look at this, could you have accomplished the same thing with another MapFeatureState?
AWAITING_ADD?

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 5]
```diff
@@ -97,68 +120,138 @@ const MapImpl: React.FunctionComponent = () => {
...
 
+      /*
+      Features in this state have been rendered. In the process of them being added to the map,
+      the variable 'needsFilter' of the appropriate type has been set to true if this was the first time a 
+      feature of that type was added to the map.
+      The step here is to check what filters need to be added and call addFilter if neccessary.
+      */
+      case MapFeatureState.RENDERED: {
+        for (const featureType of Object.values(FeatureType)) {
+          if (state.featuresByType[featureType].needsFilters) {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:174]: Rather than having boolean like this, I'd prefer to see another enum, MapFeatureTypeState.ADDED_FILTER or whatnot.

You recall the MapFeatureState enum was introduced so that there wouldn't be a bunch of booleans indicating what the state of something is.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 6]
```diff
@@ -97,68 +120,138 @@ const MapImpl: React.FunctionComponent = () => {
...
               },
             });
           }
+
           clickedUris.push(clickedFeature.uri);
         }
+        dispatch(
+          changeMapFeatureState(
+            clickedUris,
+            MapFeatureState.CLICKED_AND_LOADING
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:212]: It's not actually CLICKED_AND_LOADING if it's Transmission, though.

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:212]: Also, instead of hard-coding checks against FeatureType in multiple places, please have a function like isExpandable(featureType) or better isExpandable(feature).

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 7]
```diff
@@ -97,68 +120,138 @@ const MapImpl: React.FunctionComponent = () => {
...
+        for (const clickedFeature of featuresInState) {
+          if (latestDataCounter === limit) {
+            console.log("Made it here");
+            getFeaturesWithin({
+              variables: {
+                query: {
+                  withinFeatureUri: clickedFeature.uri,
+                },
+                limit: limit,
+                offset: totalDataCounter,
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:245]: The offset is really a per-feature state. You should track this in the Redux store. You can either (1) subclass (sub-interface) the Feature interface from the API, add a member loadingOffset?: number, and store those extended things in the Redux state or (2) keep a separate (sparse) map of feature URI -> loading state in Redux, which only has features that are currently being expanded "within". The latter is probably simpler.

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 8]
```diff
@@ -28,16 +25,25 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
+          feature.type! as keyof typeof result.featuresByType
+        ].dirty = true;
       }
       break;
     }
     case "@@kepler.gl/ADD_DATA_TO_MAP": {
       const addDataToMapAction: any = action;
       for (const row of addDataToMapAction.payload.datasets.data.rows) {
-        const addedFeature: MapFeature = row[0].properties;
+        const addedFeature: any = row[0].properties;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:45]: Don't like this..

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 9]
```diff
@@ -51,33 +57,39 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
-                  ];
-                else if (
-                  (addedFeature as any)[attribute] >
-                  filterStateOfType[attribute].max!
-                )
-                  filterStateOfType[attribute].max = (addedFeature as any)[
-                    attribute
-                  ];
-              }
+        if (!result.featureTypesFilters[addedFeature.type!]) {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:65]: You seem to be tracking multiple "per-type" states, too. You could collapse these into a single object e.g., FeatureTypeState { features; state enum (current needsFilters); max; ... } then have a map of type -> that. It's essentially your FilterState, but since filters are per-type (?) you're merging multiple things.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import {MapFeature} from "./MapFeature";
+
+export interface FeaturesByType {
+  features: MapFeature[];
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts:4]: This appears to be the FeatureTypeState I was talking about, so never mind.

File: app/geo/gui/src/ts/twxplore/gui/geo/actions/map/RepeatQueryAction.ts [hunk 11]
```diff
@@ -3,11 +3,11 @@ import {Action} from "redux-actions";
...
 
-export interface RepeatQueryAction extends Action<{uri: string}> {
+export interface RepeatQueryAction extends Action<{featureUri: string}> {
   type: REPEAT_QUERY;
 }
 
-export const repeatQuery = (uri: string): RepeatQueryAction => ({
-  payload: {uri: uri},
+export const repeatQuery = (featureUri: string): RepeatQueryAction => ({
+  payload: {featureUri: featureUri},
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/actions/map/RepeatQueryAction.ts:11]: Can use shorthand, ESLint should have flagged this.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 12]
```diff
@@ -35,6 +35,10 @@ const MapImpl: React.FunctionComponent = () => {
   const state: MapState = useSelector(
     (rootState: RootState) => rootState.app.map
   );
+  const fake_state: any = useSelector(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:38]: Should have commented this out?

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 13]
```diff
@@ -215,7 +235,10 @@ const MapImpl: React.FunctionComponent = () => {
         for (const clickedFeature of featuresInState) {
           //if the feature is expandable. Should be changed later with something like if isExpandable()
           if (clickedFeature.type !== FeatureType.Transmission) {
-            //call the lazyQuery to get feautres with the clickedFeaturee
+            for (var x = state.filterCounter - 1; x >= 0; x--) {
+              dispatch(removeFilter(x));
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:239]: I don't understand this. The filters are globally numbered?!

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 14]
```diff
@@ -294,24 +317,15 @@ const MapImpl: React.FunctionComponent = () => {
   return (
     <div>
       <div style={{width: "100%"}}>
-        <ReactResizeDetector
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:173]: Why did you remove this?

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 15]
```diff
@@ -66,82 +72,74 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
+        console.debug(
+          "changed map feature " +
+            resultFeature.uri +
+            " to state " +
+            MapFeatureState.RENDERED
+        );
+
         /*
         Check the filterState of the type of feature being added to the map.
         Loop throrugh the attributes of the feature and check to see which are of type number.
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:103]: If you've got multiple pieces of logic here, can you separate them into functions to make it more obvious? (And call them from the action handler.)

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 16]
```diff
@@ -66,82 +72,74 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
-          result.featureTypesFilters[addedFeature.type!] = {};
-          //Set featureTypeState to NEEDS_FILTERS. addFilter will be called to this type in the RENDERED case.
-          result.featuresByType[addedFeature.type!].featureTypeState =
-            MapFeatureTypeState.NEEDS_FILTERS;
-        }
-        //filterStateOfType is now the an object with properties containing relevant info needed to filter attributes of the feature type.
-        let filterStateOfType = result.featureTypesFilters[addedFeature.type!];
+
+        //filterStateOfFeatureType is now an object with properties containing relevant info needed to filter attributes of the feature type.
+        let attributesStateOfFeatureType =
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:108]: Prefer const to let.

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 17]
```diff
@@ -157,16 +155,47 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
+          );
         }
+
+        //Set the state of the feature from CLICKED_AND_LOADING to RENDERED
+        resultFeature.state = MapFeatureState.RENDERED;
+
         //Remove the loading state of the feature from loadingState
         delete result.loadingState[actionUri];
+
+        //Set FeatureTypeState of all FeatureTypes that are 'WAITING_FOR_LOAD' to 'NEEDS_FILTERS'
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:176]: Seems like a separate piece of logic, also. Separate function?

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 18]
```diff
@@ -272,22 +345,25 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
-          resultFeature.state = MapFeatureState.CLICKED;
-          console.debug(
-            "changed map feature " +
-              resultFeature.uri +
-              " to state " +
-              MapFeatureState.CLICKED
-          );
-        }
+
+      const resultFeature = result.features.find(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:349]: You're duplicating this code in several places, refactor into a function.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts [hunk 19]
```diff
@@ -1,8 +1,10 @@
 import {MapFeature} from "./MapFeature";
 import {MapFeatureTypeState} from "./MapFeatureTypeState";
+import {FilterStateTypes} from "./MapFilterStateTypes";
 
 export interface FeaturesByType {
   features: MapFeature[];
   dirty: boolean;
   featureTypeState: MapFeatureTypeState;
+  attributesState: {[attribute: string]: FilterStateTypes};
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts:9]: FilterStateTypes is really AttributeState?

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts:9]: or MapFeatureAttributeState. Please be consistent and specific about name prefixes. "Attribute" could be anything.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFilterStateTypes.ts [hunk 20]
```diff
@@ -5,5 +5,5 @@ properties on filterStateOfType
 export interface FilterStateTypes {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFilterStateTypes.ts:5]: should be same as file name, which is correct.

File: app/geo/gui/src/ts/twxplore/gui/geo/actions/map/AllFiltersSetAction.ts [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+
+export type ALL_FILTERS_SET = "ALL_FILTERS_SET";
+export const ALL_FILTERS_SET: ALL_FILTERS_SET = "ALL_FILTERS_SET";
+
+export interface AllFiltersSetAction
+  extends Action<{featureType: FeatureType}> {
+  type: ALL_FILTERS_SET;
+}
+
+export const allFiltersSet = (
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/actions/map/AllFiltersSetAction.ts:12]: Nit: usually action names start with a verb, like method names. Fine to leave this as it is.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 22]
```diff
@@ -33,7 +34,9 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
     (rootState: RootState) => rootState.app.map
   );
 
-  const featureTypeFilter = state.featuresByType[featureType].attributesState;
+  const featureTypeState = state.featuresByType[featureType].featureTypeState;
+  const featureTypeAttributeState =
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:38]: Nit: attributesState vs. AttributeState?

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 23]
```diff
@@ -47,60 +50,66 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
...
+        const idx = attributeProperties.filterIndex;
+        switch (featureTypeState) {
+          case MapFeatureTypeState.FILTERS_ADDED: {
+            //if filters have not been set yet. Attach the slider to a filter based on the attribute's unique id
+            dispatch(setFilter(idx, "name", attribute));
+            dispatch(
+              setFilter(
+                idx,
+                "type",
+                getFeatureAttributeByName(attribute).filterType
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:65]: This looks a lot nicer.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 24]
```diff
@@ -47,60 +50,66 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
...
+                    event: any,
+                    newValue: number | number[]
+                  ) => handleChange(event, newValue, attribute, idx!)}
+                  name={attribute}
+                />
+              </div>
+            );
+          }
+          default: {
+            return <React.Fragment />;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:109]: Should this ever happen? If not, throw an EvalError indicating that there's a missing case.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 25]
```diff
@@ -47,60 +50,66 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
...
+                    attributeProperties.max!,
+                  ]}
+                  getAriaValueText={valuetext}
+                  aria-labelledby="range-slider"
+                  step={1}
+                  min={attributeProperties.min!}
+                  max={attributeProperties.max!}
+                  valueLabelDisplay="auto"
+                  disabled={!attributeProperties.max!}
+                  onChangeCommitted={(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:99]: Event handlers shouldn't be inline like this, declare them as a function in the block.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 26]
```diff
@@ -24,9 +24,11 @@ import {startQuerying} from "../../actions/map/StartQueryingAction";
 import {finishLoad} from "../../actions/map/FinishLoadAction";
 import {repeatQuery} from "../../actions/map/RepeatQueryAction";
 import {MapFeatureTypeState} from "../../states/map/MapFeatureTypeState";
+import {FilterPanel} from "../filterPanel/FilterPanel";
 //import KeplerGlSchema from "kepler.gl/schemas";
 
 const limit = 500;
+const DEBUG = true;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:31]: It's fine to have this kind of constant, but I'd suggest you use an environment variable. That means you don't have to modify the source to change it.

See Environment.ts in the base lib. You could extend that class with one local to your app, and add a .debug. The values come from the webpack DefinePlugin, which can be keyed off of a command line parameter. Which would then mean you have a start-debug script in package.json.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx (hunk 22 above)
Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx]: I actually renamed this to attributeStates, because it is essentially a dictionary of several 'MapFeatureAttributeState's. 
Where key is attributeName and value is the MapFeatureAttributeState
.attributeStates[frequency] = {min: 0, max:100, idx: 5}

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 27]
```diff
@@ -33,42 +33,52 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
...
 
+  //Get the featureTypeState of the appropriate featureType which was passed in by FilterPanel as a prop
   const featureTypeState = state.featuresByType[featureType].featureTypeState;
-  const featureTypeAttributeState =
-    state.featuresByType[featureType].attributesState;
+
+  //attributeStatesOfFeatureType is now an object, with each of its properties being the state of
+  //an attribute of that FeatureType.
+  //i.e. attributeStatesOfFeature = {frequency: {min:0, max:100, filterIndex: 0}, tranmsmissionPower: {min:0, max:20, filterIndex: 1}}
+  const attributeStatesOfFeatureType =
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:45]: I like this a lot, especially with the example.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 28]
```diff
@@ -33,42 +33,52 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
...
+      {//for each attributeName string-key that points to a MapFeatureAttribute state
+      Object.keys(attributeStatesOfFeatureType).map(attributeName => {
+        const stateOfAttribute = attributeStatesOfFeatureType[attributeName]; //e.g. timestamp:{min,max}, frequency:{min, max}
+        filterIndexOfAttribute = stateOfAttribute.filterIndex;
         switch (featureTypeState) {
+          //If filters have been added
           case MapFeatureTypeState.FILTERS_ADDED: {
             //if filters have not been set yet. Attach the slider to a filter based on the attribute's unique id
-            dispatch(setFilter(idx, "name", attribute));
+            dispatch(setFilter(filterIndexOfAttribute, "name", attributeName));
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:71]: Much nicer naming here, too.

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/setAllFilterIndexNull.ts [hunk 29]
```diff
@@ -0,0 +1,10 @@
+import {MapFeatureAttributeState} from "../states/map/MapFeatureAttributeState";
+
+export const setAllFilterIndexNull = (attributeStatesOfFeatureType: {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/setAllFilterIndexNull.ts:3]: Nit: should probably be plural, FilterIndexes or FilterIndices.

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+import {getFeatureAttributeByName} from "../attributeStrategies/getFeatureAttributeByName";
+
+export const updateAttributeStatesOfFeatureType = (
+  attributeStatesOfFeatureType: {
+    [featureAttributeName: string]: MapFeatureAttributeState;
+  },
+  addedFeature: MapFeature
+) => {
+  //Loop through the feature's attribute
+  for (const attribute of Object.keys(addedFeature)) {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:12]: Nit: attributeName.

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+    //If the arttribute is numeric
+    if (getFeatureAttributeByName(attribute).isNumeric) {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:14]: Big if branch with no else. Do if (!...) { continue; }

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+      //If this is the first time coming across this attribute for this feature type
+      if (
+        !attributeStatesOfFeatureType[attribute].min &&
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:17]: Duplicate expressions?

Also, min could be 0. You probably mean to check typeof(attributeStatesOfFeatureType[attribute].min) === "undefined".

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts [hunk 33]
```diff
@@ -0,0 +1,50 @@
...
+        !attributeStatesOfFeatureType[attribute].min
+      ) {
+        //New max of the attribute wil be the feature's value for the attribute
+        attributeStatesOfFeatureType[attribute].max = (addedFeature as any)[
+          attribute
+        ];
+        //New min of the attribute wil be the feature's value for the attribute
+        attributeStatesOfFeatureType[attribute].min = (addedFeature as any)[
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:25]: Not sure about this addedFeature as any business.

I'd rather see something like:

const attributeKey = attributeName as keyof MapFeatureType.

(https://github.com/microsoft/TypeScript/pull/12253#issuecomment-263132208)

Then you can dispense with this as any business.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts [hunk 34]
```diff
@@ -1,10 +1,10 @@
...
 import {MapFeatureTypeState} from "./MapFeatureTypeState";
-import {FilterStateTypes} from "./MapFilterStateTypes";
+import {MapFeatureAttributeState} from "./MapFeatureAttributeState";
 
 export interface FeaturesByType {
   features: MapFeature[];
   dirty: boolean;
   featureTypeState: MapFeatureTypeState;
-  attributesState: {[attribute: string]: FilterStateTypes};
+  attributeStates: {[featureAttributeName: string]: MapFeatureAttributeState};
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts:9]: I'm not sure TS allows this, but perhaps this could be [keyof MapFeature]

Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/states/map/FeaturesByType.ts:9]: Didn't allow 'Union-types' to be used as an index unfortunately

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx [hunk 35]
```diff
@@ -53,16 +53,16 @@ const FilterSlidersImpl: React.FunctionComponent<{featureType: string}> = ({
   ) => {
     dispatch(setFilter(filterIndexOfAttribute, "value", newValue));
   };
-  const onChangeCommitted = (event: any, newValue: number | number[]) => {
-    handleChange(event, newValue, filterIndexOfAttribute!);
-  };
+  //const onChangeCommitte = (event: any, newValue: number | number[]) => {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterSliders.tsx:56]: Nit: I usually don't leave non-functional commented-out code in like this.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 36]
```diff
@@ -185,9 +186,12 @@ const MapImpl: React.FunctionComponent = () => {
         //loop through each feature type
         for (const featureType of Object.values(FeatureType)) {
           //Check if filters need to be added for this FeatureType
+          //We don't want to addFilters when some featuresByType lists are
+          //still dirty because then the filters will be removed by removeDataset() in the LOADED case
           if (
             state.featuresByType[featureType].featureTypeState ===
-            MapFeatureTypeState.NEEDS_FILTERS
+              MapFeatureTypeState.NEEDS_FILTERS &&
+            noDirtyFeaturesListCheck(state.featuresByType)
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:194]: Nit: that's a weird name for a function. hasDirtyFeatures ?

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:194]: You're only using this function in one place. Just declare it inline or as a top-level function (not exported) in Map.tsx?

In any case, this function should really be a one-liner, using Array.find() or the lodash (_).find. Something like:
!_.find(featuresByType, features => features.dirty).
https://lodash.com/docs/2.4.2#find

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts [hunk 37]
```diff
@@ -9,42 +9,42 @@ export const updateAttributeStatesOfFeatureType = (
...
+  for (const attributeName of Object.keys(addedFeature)) {
+    const attributeKey = attributeName as keyof MapFeature;
     //If the arttribute is numeric
-    if (getFeatureAttributeByName(attribute).isNumeric) {
+    if (getFeatureAttributeByName(attributeName).isNumeric) {
       //If this is the first time coming across this attribute for this feature type
       if (
-        !attributeStatesOfFeatureType[attribute].min &&
-        !attributeStatesOfFeatureType[attribute].min
+        !attributeStatesOfFeatureType[attributeName].min &&
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:18]: Still not handling the case where min and/or max could be 0. Should be typeof(...) === "undefined".

Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:18]: Seems like you could put attributeStatesOfFeatureType[attributeName] into a variable, or?

Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:18]: I initialize all possible combinations of  attributeStatesOfFeaatureType[attribute].min/max to null in initialMapStates.tsx so it can never undefined.
If min/max is 0 then that means this is not the first time coming across the attribute, because a feature must have had 0 for transmissionPower for example, and that's what changed min/max to 0

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:18]: My point remains. !0 and !null will both be true in this branch.

Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts]: Variable named attributeStateOfFeatureType.
Which is awfully close to attributesStatesOfFeatureType[attributeName] but it's an accurate name.

Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts]: Checked for null by using === equality check instead of !attributeStates...
